rfcs/rfc-111.md

```diff
@@ -0,0 +1,354 @@
+
+Since releasing guard in October of 2020, the CloudFormation team has gathered feedback from a variety of sources on the current rule language authoring experience. From this feedback, we are proposing some enhancements to the existing ruleset language in the form of grouped type and rule blocks to address the below issues.
+
+## Tenets (unless you know better ones)
+**Simple**: The language must be simple for customers to author rules, simple for IDE integrations, readable for human comprehension while being machine enforceable.
+
+**Unambiguous**: The language must not allow for ambiguous interpretations that makes it hard to comprehend what is being evaluated. The tool is targeted for security and compliance related attestations that need the auditor to consistent and unambiguously understand rules and their evaluations.
+
+**Deterministic**: The language design must allow language implementers to have deterministic, consistent and isolated  evalutions. Results for repeated evaluations for the same context and rule set must evaluate to the same result everytime. Time to evaluate results inside near identical environments must be within acceptable tolerance limits.
+
+**Composable**: The language makes composition of higher order rule sets from multiple different rules sets simple with consistent interpretation and syntax. Composition should not add complexity to interpretation and customers can easily navigate across them.
+
+## Issues with Current Language
+
+Below are the issues we are attempting to address with the introduction of block evaluations, along with samples of the proposed enhancements to the language.
+
+### **Increased verbosity**
+
+Current rulesets are extremely verbose, with information required to be repeated for every rule, as seen in the example below:
+
+```
+AWS::ApiGateway::Method AuthorizationType == NONE
+AWS::ApiGateway::Method HttpMethod == ANY
+AWS::ApiGateway::Method Integration == {"IntegrationHttpMethod":"POST","Type":"AWS_PROXY","Uri":"arn:aws:apigateway:${AWS::Region}:lambda:path/2015-03-31/functions/${LambdaWAFBadBotParserFunction.Arn}/invocations"}
+AWS::ApiGateway::Method RequestParameters == {"method.request.header.X-Forwarded-For":false}
+AWS::ApiGateway::Method ResourceId == ApiGatewayBadBot.RootResourceId |OR| AWS::ApiGateway::Method ResourceId == ApiGatewayBadBotResource
+AWS::ApiGateway::Method RestApiId == ApiGatewayBadBot
+```
+
+In the above ruleset, every statement clause needs to repeat the type information like AWS::ApiGateway::Method, type has to be repeated across OR clauses, and OR clauses have to present on a single line which makes it hard to read end to end. This violates our tenet of simplicity. The current rule authoring experience is quite tedious because of this, with a large amount of information replicated line to line that is the same for a given resource type.
```

Comment:
[copied from our previous discussion] think there's multiple sides to this. Lines being their own rules that can be evaluated independently and being able to sort and autocomplete rules more easily is arguably simpler